File: src/services/productApi.ts
import { apiClient } from "./apiClient";
import type { Product } from "../types/product";

type ApiProduct = Omit<Product, "stock"> & {
  stock?: number;
  stockQuantity?: number;
};

type ProductsResponse = ApiProduct[] | { products?: ApiProduct[] };

function unwrap<T>(response: { data: { data: T } }): T {
  return response.data.data;
}

function normalizeProduct(product: ApiProduct): Product {
  return {
    ...product,
    stock: product.stock ?? product.stockQuantity ?? 0,
  };
}

export async function getProducts(): Promise<Product[]> {
  const payload = unwrap<ProductsResponse>(await apiClient.get("/products"));
  const products = Array.isArray(payload) ? payload : payload.products;
  return Array.isArray(products) ? products.map(normalizeProduct) : [];
}

export async function getProductById(id: string): Promise<Product | undefined> {
  const product = unwrap<ApiProduct>(await apiClient.get(`/products/${id}`));
  return normalizeProduct(product);
}

export async function createProduct(
  data: Omit<Product, "id" | "createdAt" | "updatedAt">,
): Promise<Product> {
  const product = unwrap<ApiProduct>(
    await apiClient.post("/products", {
      ...data,
      stockQuantity: data.stock,
    }),
  );
  return normalizeProduct(product);
}

export async function updateProduct(
  id: string,
  data: Partial<Omit<Product, "id" | "createdAt">>,
): Promise<Product | null> {
  const product = unwrap<ApiProduct>(
    await apiClient.put(`/products/${id}`, {
      ...data,
      ...(data.stock !== undefined ? { stockQuantity: data.stock } : {}),
    }),
  );
  return normalizeProduct(product);
}

export async function deleteProduct(id: string): Promise<boolean> {
  await apiClient.delete(`/products/${id}`);
  return true;
}

File: mobile-flutter/lib/core/api/product_api.dart
import '../models/product.dart';
import 'api_client.dart';

class ProductApi {
  final ApiClient _client;
  ProductApi(this._client);

  Future<List<Product>> getProducts() async {
    final data = await _client.get('/products');
    if (data is List) {
      return data
          .map((e) => Product.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    if (data is Map) {
      final inner = data['products'] ?? data['data'];
      if (inner is List) {
        return inner
            .map((e) => Product.fromJson(e as Map<String, dynamic>))
            .toList();
      }
    }
    return [];
  }

  Future<Product> getProduct(String id) async {
    final data = await _client.get('/products/$id');
    return Product.fromJson(data as Map<String, dynamic>);
  }

  Future<Product> createProduct(Map<String, dynamic> payload) async {
    final data = await _client.post('/products', payload);
    return Product.fromJson(data as Map<String, dynamic>);
  }

  Future<Product> updateProduct(String id, Map<String, dynamic> payload) async {
    final data = await _client.put('/products/$id', payload);
    return Product.fromJson(data as Map<String, dynamic>);
  }

  Future<void> deleteProduct(String id) async {
    await _client.delete('/products/$id');
  }
}
